scala/runtime/src/test/scala/org/coursera/courier/templates/CompanionRaceTest.scala

```diff
@@ -0,0 +1,157 @@
+import java.nio.ByteBuffer
+
+import scala.io.Source
+import sun.misc.URLClassPath
+
+class TestBridge {
+  def value = TestCompanion.VALUE.toString
+  def withName = TestCompanion.withName("VALUE").toString
+}
+
+/**
+  * This test shows that the deadlock tested for in [[EnumTemplateRaceTest]]
+  * can be reproduced with relatively simple scala only, with no dependence
+  * on courier or courscala libraries. This deadlock uses two lazy values and
+  * a companion class initialization lock.
+  */
+class CompanionRaceTest() {
+
+  /**
+    * Make a separate class loader for a test, and load the [[TestBridge]] class.
+    * The bug occurs while loading classes, hence the need for a class loader per test.
+    * @return The TestBridge class in the new class loader.
+    */
+  def createForeignClazz(): Class[_] = {
+    val systemClassLoader = this.getClass.getClassLoader()
+
+    val classLoader = new ClassLoader(systemClassLoader.getParent) {
+      override def findClass(name: String): Class[_] = {
+        val path = name.replace('.', '/').concat(".class")
+        val resOpt = systemClassLoader.getResourceAsStream(path)
+        Option(resOpt) match {
+          case Some(res) =>
+            try {
+              val bytes = Stream.continually(res.read).takeWhile(_ != -1).map(_.toByte).toArray
+              defineClass(name, ByteBuffer.wrap(bytes), null)
+            } catch {
+              case e: IOException => throw new ClassNotFoundException(name, e)
+            }
+          case None => throw new ClassNotFoundException(name)
+        }
+      }
+      override def loadClass(name: String, resolve: Boolean) =
+        super.loadClass(name, true)
+    }
+    classLoader.loadClass(new TestBridge().getClass.getName)
+  }
+
+  class TestSetup() extends Race {
+
+    private val bridgeClazz = createForeignClazz()
+    def method(name: String) = bridgeClazz.getMethod(name)
+    private var foreignObject = bridgeClazz.getConstructor().newInstance()
+
+    def changeInstance(): Unit = {
+      foreignObject = bridgeClazz.getConstructor().newInstance()
+    }
+
+    def value(): String = s"""${method("value").invoke(foreignObject)}"""
+    def withName(): String = s"""${method("withName").invoke(foreignObject)}"""
+
+  }
+  @Test
+  def valueBeforeRace(): Unit = {
+    val testSetup = new TestSetup()
+    testSetup.value()
+    testSetup.changeInstance()
+    assert(testSetup.race(testSetup.value(), testSetup.withName()))
+  }
+  @Test
+  def raceToDeadlock(): Unit = {
+    val testSetup = new TestSetup()
+    assert(testSetup.race(testSetup.value(), testSetup.withName()))
+  }
+
+  /**
+    * This is [[CompanionRaceTest.raceToDeadlock]] modified to avoid the race
+    * by running one of the halves first, which initializes all the classes,
+    * we then change the instance in `testSetup` and run the race, which does
+    * not deadlock - showing that the class initialization lock is involved in
+    * the deadlock.
+    */
```

Comment:
This is very clever. I can follow along with some effort, but I don't think I would have come up with the series of refinements that trace this down to class initialization.